python/lsst/skymap/patchInfo.py

```diff
@@ -22,62 +22,105 @@
 
 __all__ = ["PatchInfo", "makeSkyPolygonFromBBox"]
 
-from lsst.sphgeom import ConvexPolygon
-from lsst.geom import Box2D
+import numbers
+from collections.abc import Iterable
 
-
-def makeSkyPolygonFromBBox(bbox, wcs):
-    """Make an on-sky polygon from a bbox and a SkyWcs
-
-    Parameters
-    ----------
-    bbox : `lsst.geom.Box2I` or `lsst.geom.Box2D`
-        Bounding box of region, in pixel coordinates
-    wcs : `lsst.afw.geom.SkyWcs`
-        Celestial WCS
-
-    Returns
-    -------
-    polygon : `lsst.sphgeom.ConvexPolygon`
-        On-sky region
-    """
-    pixelPoints = Box2D(bbox).getCorners()
-    skyPoints = wcs.pixelToSky(pixelPoints)
-    return ConvexPolygon.convexHull([sp.getVector() for sp in skyPoints])
+from lsst.geom import Extent2I, Point2I, Box2I
+from .detail import makeSkyPolygonFromBBox, Index2D
+from .cellInfo import CellInfo
 
 
 class PatchInfo:
     """Information about a patch within a tract of a sky map.
 
+    If cellInnerDimensions and cellBorder are set then the patch
+    will be gridded with cells.
+
     See `TractInfo` for more information.
 
     Parameters
     ----------
-    index : `tuple` of `int`
+    index : `lsst.skymap.Index2D`
         x,y index of patch (a pair of ints)
     innerBBox : `lsst.geom.Box2I`
         inner bounding box
     outerBBox : `lsst.geom.Box2I`
         inner bounding box
+    sequentialIndex : `int`
+        Patch sequential index
+    tractWcs : `lsst.afw.geom.SkyWcs`
+        Tract WCS object.
+    cellInnerDimensions : `Iterable` [`int`, `int`] or `lsst.geom.Extent2I`, optional
+        Inner dimensions of each cell (x,y pixels).
+    cellBorder : `int`, optional
+        Cell border size (pixels).
+    numCellsPerPatchInner : `int`, optional
+        Number of cells per inner patch region.
+    numCellsInPatchBorder : `int`, optional
+        Number of cells in the patch border.
     """
 
-    def __init__(self, index, innerBBox, outerBBox):
+    def __init__(self, index, innerBBox, outerBBox, sequentialIndex,
+                 tractWcs=None,
+                 cellInnerDimensions=(0, 0), cellBorder=0,
+                 numCellsPerPatchInner=0, numCellsInPatchBorder=0):
         self._index = index
+        self._sequentialIndex = sequentialIndex
         self._innerBBox = innerBBox
         self._outerBBox = outerBBox
+        self._wcs = tractWcs
```

Comment:
I'm not sure what the comment is here?

Reply:
Let me try to rephrase my comment.

Since `PatchInfo` can be instantiated with `tractWCS` set to `None` (default), `self._wcs` will be set to `None`. I don't see a way where `_wcs` can be set later after instantiation. This would cause calling`getInnerSkyPolygon` and `getOuterSkyPolygon` with `tractWCS=None` fail (default again). So we should not allow a `PatchInfo` object to be constructed without a `tractWCS`.

I recollect discussing about overriding the WCS attached to the object when calling those methods, but don't recollect discussing about needing to create `PatchInfo` without a WCS. If you see a compelling need, then please add to the docstring that if initialized without a WCS, they must necessarily be passed in methods such as `getInnerSkyPolygon`.

Reply:
Oh, I see your point.  I _think_ that there's no way for anything legitimate (not monkey-patched or otherwise hacked) to get a `PatchInfo` without a wcs.  So I could make this required at instantiation, or at least Raise if it doesn't have a valid wcs.  Which would you prefer?

Reply:
I would prefer making it required instead of raising an exception. I could imagine trying to create a PatchInfo directly for testing/playing around and would like to know that I need to provide a WCS to begin with.